### ISSUE TITLE
12.0 Bank Reconciliation Widget is not multicompany safe

### ISSUE THREAD

**Issue by gdgellatly** [odoo/odoo#32954]
Impacted versions: 12.0

Steps to reproduce:
Have 2 companies A & B
Do not share partners
Have a partner that buys from both companies (actually the live case is 2 different partners that pay from same bank account in different comanies).
Use bank statement imports, such that the bank account number is set.
After import click reconcile.
Boom

The function 
```python
    @api.model
    def _get_bank_statement_line_partners(self, st_lines):
        query = '''
            SELECT
                st_line.id                          AS id,
                COALESCE(p1.id,p2.id,p3.id)         AS partner_id
            FROM account_bank_statement_line st_line
            LEFT JOIN res_partner_bank bank         ON bank.id = st_line.bank_account_id OR bank.acc_number = st_line.account_number
            LEFT JOIN res_partner p1 ON st_line.partner_id=p1.id
            LEFT JOIN res_partner p2 ON bank.partner_id=p2.id
            LEFT JOIN res_partner p3 ON p3.name ILIKE st_line.partner_name
            WHERE st_line.id IN %s
        '''
        params = [tuple(st_lines.ids)]

        # Add the res.partner's IR rules to the WHERE clause. In case partners are not shared
        # between companies, identical partners may exist in a company we don't have access to.
        ir_rules_query = self.env['res.partner']._where_calc([])
        self.env['res.partner']._apply_ir_rules(ir_rules_query, 'read')
        from_clause, where_clause, where_clause_params = ir_rules_query.get_sql()
        where_p3 = (" AND %s" % where_clause).replace('res_partner', 'p3') if where_clause else ''
        if where_p3:
            query += where_p3
            params += where_clause_params

        self._cr.execute(query, params)

        result = {}
        for res in self._cr.dictfetchall():
            result[res['id']] = res['partner_id']
        return result
```

has multiple flaws. It only checks rules for p3. Yet all of res.partner.bank and p1 and p2 are also company constrained.

Current behavior:
Access Error as wrong partner is picked up from prior import in another company, (p2)

Expected behavior:
If using SQL to display data to user, actually enforce record rules and ACL's same as ORM.

Video/Screenshot link (optional):

**Comment by gdgellatly** [comment]
opw 1974649

**Comment by sswapnesh** [comment]
> If using SQL to display data to user, actually enforce record rules and ACL's same as ORM.

That what Odoo does not support on SQL

**Comment by gdgellatly** [comment]
@sswapnesh Of course it supports it. That is precisely what building where_p3 does. Just not done for the other models.

**Comment by nim-odoo** [comment]
Hello,

> Have a partner that buys from both companies

That's not possible if partners are not shared, unless you:
- share the partners
- invoice in 2 companies
- do not share the partners

Which doesn't seem to be a valid use case.

> 2 different partners that pay from same bank account in different comanies

That one I can understand it, but I can't reproduce it. Could you please provide an example of statement file? By the way, I think the issue comes from: `OR bank.acc_number = st_line.account_number`

There is an attempt to add the IR rule at https://github.com/odoo/odoo/pull/33245/files#diff-34f8ae8a5209bc49b57e881ee41799d7R105 (I think the modification above is useless). However, since I can't reproduce, I can't test.

**Comment by nim-odoo** [comment]
@gdgellatly It should be fixed with #33245 Let me know what you think about it.

**Comment by sswapnesh** [comment]
> > If using SQL to display data to user, actually enforce record rules and ACL's same as ORM.
> 
> That what Odoo does not support on SQL

Mybad, I was talking in general terms not this one.
